lib/utils.js

```diff
@@ -39,6 +39,21 @@ exports.pick = function (obj, arr) {
   return res;
 };
 
+// just like emitter.once(event, listener), but doesn't
+// add additional listeners if any listeners with the
+// key are already attached
+exports.oncePerKey = function oncePerKey(emitter, event, key, listener) {
+  var listeners = emitter.listeners(event).filter(function (listener) {
+    return listener._key === key;
+  });
+  if (listeners.length) {
+    return;
+  }
+  emitter.once(event, listener);
+  listeners = emitter.listeners(event);
+  listeners[listeners.length - 1]._key = key;
+};
+
```

Comment:
I could also make `oncePerKey` remove the old listeners and add the new one. Should be 6 of one, half a dozen of the other.

In general, the way I would like to solve this in the future is for `retry` replication to just use `live` replication's APIs &ndash; i.e. to be implemented on top of it, rather than inside of it.

In any case, I don't think this objection is enough to block this PR. Sure the original listeners will maintain references to the original closure, but at least the listeners don't grow out of control. :)